File: bindings/java/src/io/quad4/rns/Event.java
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) 2024-2026 Quad4.io

package io.quad4.rns;

import com.sun.jna.Memory;
import io.quad4.rns.ffi.RnsLibrary;

/** Polled librns event. */
public final class Event {
    public static final int ANNOUNCE = RnsLibrary.RNS_EV_ANNOUNCE;
    public static final int LINK_ESTABLISHED = RnsLibrary.RNS_EV_LINK_ESTABLISHED;
    public static final int LINK_FAILED = RnsLibrary.RNS_EV_LINK_FAILED;
    public static final int LINK_DATA = RnsLibrary.RNS_EV_LINK_DATA;
    public static final int LINK_CLOSED = RnsLibrary.RNS_EV_LINK_CLOSED;
    public static final int REQUEST_INCOMING = RnsLibrary.RNS_EV_REQUEST_INCOMING;
    public static final int REQUEST_RESPONSE = RnsLibrary.RNS_EV_REQUEST_RESPONSE;
    public static final int REQUEST_FAILED = RnsLibrary.RNS_EV_REQUEST_FAILED;
    public static final int RESOURCE_STARTED = RnsLibrary.RNS_EV_RESOURCE_STARTED;
    public static final int RESOURCE_CONCLUDED = RnsLibrary.RNS_EV_RESOURCE_CONCLUDED;
    public static final int DESTINATION_DATA = RnsLibrary.RNS_EV_DESTINATION_DATA;

    private final int kind;
    private final int hops;
    private final byte[] linkId;
    private final byte[] destinationHash;
    private final byte[] identityHash;
    private final byte[] requestId;
    private final String path;
    private final String errorMessage;
    private final byte[] appData;
    private final boolean appDataTruncated;

    private Event(
            int kind,
            int hops,
            byte[] linkId,
            byte[] destinationHash,
            byte[] identityHash,
            byte[] requestId,
            String path,
            String errorMessage,
            byte[] appData,
            boolean appDataTruncated) {
        this.kind = kind;
        this.hops = hops;
        this.linkId = linkId;
        this.destinationHash = destinationHash;
        this.identityHash = identityHash;
        this.requestId = requestId;
        this.path = path;
        this.errorMessage = errorMessage;
        this.appData = appData;
        this.appDataTruncated = appDataTruncated;
    }

    public static Event poll(Node node, int timeoutMs, int appDataCapacity) {
        RnsLibrary.RnsEvent raw = new RnsLibrary.RnsEvent();
        Memory storage = null;
        if (appDataCapacity > 0) {
            storage = new Memory(appDataCapacity);
            raw.app_data = storage;
            raw.app_data_cap = appDataCapacity;
        }
        int code = RnsLibrary.INSTANCE.rns_event_poll(node.handle(), raw, timeoutMs);
        Rns.check(code);

        byte[] data = new byte[0];
        if (raw.app_data_len > 0 && storage != null) {
            data = storage.getByteArray(0, (int) raw.app_data_len);
        }

        return new Event(
                raw.kind,
                raw.hops & 0xff,
                RnsLibrary.copyHash(raw.link_id, raw.link_id_len),
                RnsLibrary.copyHash(raw.destination_hash, raw.destination_hash_len),
                RnsLibrary.copyHash(raw.identity_hash, raw.identity_hash_len),
                RnsLibrary.copyHash(raw.request_id, raw.request_id_len),
                RnsLibrary.cstr(raw.path),
                RnsLibrary.cstr(raw.error_message),
                data,
                raw.app_data_truncated != 0);
    }

    public int kind() {
        return kind;
    }

    public int hops() {
        return hops;
    }

    public byte[] linkId() {
        return linkId.clone();
    }

    public byte[] destinationHash() {
        return destinationHash.clone();
    }

    public byte[] identityHash() {
        return identityHash.clone();
    }

    public byte[] requestId() {
        return requestId.clone();
    }

    public String path() {
        return path;
    }

    public String errorMessage() {
        return errorMessage;
    }

    public byte[] appData() {
        return appData.clone();
    }

    public boolean appDataTruncated() {
        return appDataTruncated;
    }
}

File: bindings/lua/rns/event.lua
-- SPDX-License-Identifier: Apache-2.0
-- Copyright (c) 2024-2026 Quad4.io

local ffi_mod = require("rns.ffi")
local errors = require("rns.errors")
local ffi = ffi_mod.ffi
local lib = ffi_mod.lib

local EventKind = {
	NONE = 0,
	ANNOUNCE = 1,
	LINK_ESTABLISHED = 2,
	LINK_FAILED = 3,
	LINK_DATA = 4,
	LINK_CLOSED = 5,
	REQUEST_INCOMING = 6,
	REQUEST_RESPONSE = 7,
	REQUEST_FAILED = 8,
	RESOURCE_STARTED = 9,
	RESOURCE_CONCLUDED = 10,
	DESTINATION_DATA = 11,
}

local Event = {}
Event.__index = Event

function Event.poll(node, timeout_ms, app_data_cap)
	local event = ffi.new("rns_event")
	local buf = nil
	local cap = app_data_cap or 0
	if cap > 0 then
		buf = ffi.new("uint8_t[?]", cap)
		event.app_data = buf
		event.app_data_cap = cap
	end
	errors.map_code(lib.rns_event_poll(node:handle(), event, timeout_ms or 0))
	local data = ""
	if event.app_data ~= nil and event.app_data_len > 0 then
		data = ffi.string(event.app_data, event.app_data_len)
	end
	return setmetatable({ _raw = event, _app_data = data, _keep = buf }, Event)
end

function Event:kind()
	return tonumber(self._raw.kind)
end

function Event:hops()
	return tonumber(self._raw.hops)
end

function Event:link_id()
	return ffi.string(self._raw.link_id, self._raw.link_id_len)
end

function Event:destination_hash()
	return ffi.string(self._raw.destination_hash, self._raw.destination_hash_len)
end

function Event:identity_hash()
	return ffi.string(self._raw.identity_hash, self._raw.identity_hash_len)
end

function Event:request_id()
	return ffi.string(self._raw.request_id, self._raw.request_id_len)
end

function Event:path()
	return ffi_mod.cstr(self._raw.path, 256)
end

function Event:error_message()
	return ffi_mod.cstr(self._raw.error_message, 256)
end

function Event:app_data()
	return self._app_data
end

function Event:app_data_truncated()
	return self._raw.app_data_truncated ~= 0
end

local function hash_eq(a, b)
	return type(a) == "string" and type(b) == "string"
		and #a == ffi_mod.HASH_LEN and #b == ffi_mod.HASH_LEN and a == b
end

return {
	Event = Event,
	EventKind = EventKind,
	hash_eq = hash_eq,
}
